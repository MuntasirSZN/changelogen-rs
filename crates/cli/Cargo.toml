[package]
name = "novalyn"
version.workspace = true
authors.workspace = true
edition.workspace = true
rust-version.workspace = true
description = "ðŸ’… Beautiful Changelogs using Conventional Commits"
readme.workspace = true
homepage.workspace = true
repository.workspace = true
license.workspace = true
keywords.workspace = true
categories.workspace = true

[package.metadata.wix]
upgrade-guid = "E71412B2-B395-4566-9FB6-53EBCC3526DF"
path-guid = "5A0D56D6-30D8-41CE-B5F6-A07E3BB98D6A"
license = false
eula = false

[dependencies]
anyhow.workspace = true
clap.workspace = true
clap_complete.workspace = true
clap_complete_nushell.workspace = true
novalyn_core = { path = "../core", version = "0.1.0" }
tokio.workspace = true
tracing-subscriber.workspace = true

[dev-dependencies]
assert_cmd.workspace = true
assert_fs.workspace = true
insta.workspace = true
predicates.workspace = true

[build-dependencies]
clap.workspace = true
clap_complete.workspace = true
clap_complete_nushell.workspace = true
clap_mangen.workspace = true

[target.'cfg(not(target_env = "msvc"))'.dependencies]
mimalloc-safe.workspace = true

[target.'cfg(target_arch = "aarch64")'.dependencies]
mimalloc-safe = { workspace = true, features = [
  "no_opt_arch",
] }

[target.'cfg(target_env = "msvc")'.dependencies]
mimalloc-safe = { workspace = true, features = [
  "override",
] }
