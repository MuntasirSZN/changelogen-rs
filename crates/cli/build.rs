use clap::CommandFactory;
use clap_complete::generate_to;
use std::env::var;
use std::fs;

include!("src/shells.rs");
include!("src/cli_def.rs");

fn main() -> std::io::Result<()> {
    let mut cmd = Cli::command();
    let out_dir = var("OUT_DIR").unwrap();
    let completion_out_dir = Path::new(&out_dir).join("completions");

    fs::create_dir_all(&completion_out_dir)?;

    for shell in Shell::value_variants() {
        generate_to(*shell, &mut cmd, "novalyn", &completion_out_dir)?;
    }

    let man_out_dir = Path::new(&out_dir).join("man");
    fs::create_dir_all(&man_out_dir)?;

    // Generate man pages for main command and all subcommands
    let cmd = Cli::command();
    clap_mangen::generate_to(cmd, &man_out_dir)?;

    Ok(())
}
