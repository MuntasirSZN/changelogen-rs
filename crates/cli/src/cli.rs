use crate::{
    github, logging,
    pipeline::{ExitCode, ReleaseOptions, run_release},
};
use anyhow::Result;
use clap::{CommandFactory, Parser};
use clap_complete;
use ecow::EcoVec;

pub use crate::cli_def::{Cli, Commands, Completions};

pub fn run() -> Result<ExitCode> {
    let cli = Cli::parse();
    logging::init(cli.verbose as usize);
    let cwd = cli
        .cwd
        .as_ref()
        .map(std::path::PathBuf::from)
        .unwrap_or(std::env::current_dir()?);
    let exit = match cli.command {
        Commands::Completions(completions) => {
            let mut cmd = Cli::command();
            clap_complete::generate(
                completions.shell,
                &mut cmd,
                "novalyn",
                &mut std::io::stdout(),
            );
            ExitCode::Success
        }
        Commands::Show {
            from,
            to,
            new_version,
        } => {
            let parsed_new = new_version.and_then(|s| semver::Version::parse(&s).ok());
            let outcome = run_release(ReleaseOptions {
                cwd,
                from: from.map(|s| s.into()),
                to: to.map(|s| s.into()),
                dry_run: true,
                new_version: parsed_new,
                no_authors: true,
                exclude_authors: EcoVec::new(),
                hide_author_email: false,
                clean: false,
                sign: false,
                yes: true, // Show command doesn't need confirmation
                github_alias: false,
                github_token: None,
            })?;
            println!("{}", outcome.version);
            ExitCode::Success
        }
        Commands::Generate {
            write,
            output,
            from,
            to,
            new_version,
            no_authors,
            exclude_author,
            hide_author_email,
            clean,
            sign,
            yes,
            no_github_alias,
            github_token,
        } => {
            // Read GitHub token from env if not provided
            let github_token = github_token.or_else(|| {
                std::env::var("GITHUB_TOKEN")
                    .ok()
                    .or_else(|| std::env::var("GH_TOKEN").ok())
            });

            let parsed_new = new_version.and_then(|s| semver::Version::parse(&s).ok());
            let outcome = run_release(ReleaseOptions {
                cwd: cwd.clone(),
                from: from.map(|s| s.into()),
                to: to.map(|s| s.into()),
                dry_run: !write,
                new_version: parsed_new,
                no_authors,
                exclude_authors: exclude_author.into_iter().map(|s| s.into()).collect(),
                hide_author_email,
                clean,
                sign,
                yes,
                github_alias: !no_github_alias,
                github_token: github_token.map(|s| s.into()),
            })?;
            if let Some(path) = output {
                std::fs::write(&path, outcome.version.to_string())?;
            }
            println!(
                "Generated v{} ({} commits){}",
                outcome.version,
                outcome.commit_count,
                if write {
                    if outcome.wrote {
                        " and updated CHANGELOG.md"
                    } else {
                        " (no change)"
                    }
                } else {
                    ""
                }
            );
            if !outcome.wrote && write {
                ExitCode::NoChange
            } else {
                ExitCode::Success
            }
        }
        Commands::Release {
            dry_run,
            from,
            to,
            new_version,
            no_authors,
            exclude_author,
            hide_author_email,
            clean,
            sign,
            yes,
            no_github_alias,
            github_token,
        } => {
            // Read GitHub token from env if not provided
            let github_token = github_token.or_else(|| {
                std::env::var("GITHUB_TOKEN")
                    .ok()
                    .or_else(|| std::env::var("GH_TOKEN").ok())
            });

            let parsed_new = new_version.and_then(|s| semver::Version::parse(&s).ok());
            let outcome = run_release(ReleaseOptions {
                cwd: cwd.clone(),
                from: from.map(|s| s.into()),
                to: to.map(|s| s.into()),
                dry_run,
                new_version: parsed_new,
                no_authors,
                exclude_authors: exclude_author.into_iter().map(|s| s.into()).collect(),
                hide_author_email,
                clean,
                sign,
                yes,
                github_alias: !no_github_alias,
                github_token: github_token.map(|s| s.into()),
            })?;
            if outcome.wrote {
                println!("Released v{}", outcome.version);
                ExitCode::Success
            } else {
                println!("No change for v{}", outcome.version);
                ExitCode::NoChange
            }
        }
        Commands::Github { tag, body_path } => {
            // Minimal body read
            let body = if let Some(path) = body_path {
                std::fs::read_to_string(path)?
            } else {
                String::new()
            };
            // attempt repo detection via config layer
            let cfg = crate::config::load_config(crate::config::LoadOptions {
                cwd: &cwd,
                cli_overrides: None,
            })?;
            if let Some(repo) = cfg.repo {
                let rt = tokio::runtime::Runtime::new()?;
                let info = rt.block_on(async move {
                    github::sync_release(&repo, cfg.github_token.as_deref(), &tag, &body, None)
                        .await
                });
                match info {
                    Ok(r) => {
                        println!(
                            "GitHub release {}: {} (created={}, updated={}, skipped={})",
                            r.tag, r.url, r.created, r.updated, r.skipped
                        );
                        ExitCode::Success
                    }
                    Err(e) => {
                        eprintln!("github sync error: {e}");
                        ExitCode::NoChange
                    }
                }
            } else {
                eprintln!("no repository info available");
                ExitCode::NoChange
            }
        }
    };
    Ok(exit)
}
#[cfg(test)]
mod tests {
    use super::*;
    use clap::CommandFactory;
    #[test]
    fn cli_help_generates() {
        Cli::command().debug_assert();
    }
}
