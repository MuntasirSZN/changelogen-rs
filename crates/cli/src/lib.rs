//! novalyn CLI crate entry point
//!
//! This crate provides the command-line interface for novalyn, a parity-focused changelog generator.
#![forbid(unsafe_code)]

pub mod cli;
pub mod cli_def;
pub mod logging;
pub mod shells;

pub use novalyn_core::*;
