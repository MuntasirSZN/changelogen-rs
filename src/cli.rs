use crate::{
    github, logging,
    pipeline::{ExitCode, ReleaseOptions, run_release},
};
use anyhow::Result;
use clap::{ArgAction, Parser, Subcommand};

#[derive(Parser, Debug)]
#[command(
    name = "changelogen",
    version,
    about = "Generate changelogs from conventional commits."
)]
pub struct Cli {
    #[command(subcommand)]
    pub command: Commands,
    #[arg(long)]
    pub cwd: Option<String>,
    /// Increase verbosity (-v, -vv, -vvv)
    #[arg(short = 'v', action = ArgAction::Count)]
    pub verbose: u8,
}

#[derive(Subcommand, Debug)]
pub enum Commands {
    /// Show the next inferred version
    Show {
        #[arg(long)]
        from: Option<String>,
        #[arg(long)]
        to: Option<String>,
        #[arg(long, value_name = "SEMVER")]
        new_version: Option<String>,
    },
    /// Generate release block (writes with --write)
    Generate {
        #[arg(long)]
        write: bool,
        /// Write output to file instead of stdout (implies --write when path is CHANGELOG.md)
        #[arg(long, value_name = "PATH")]
        output: Option<String>,
        #[arg(long)]
        from: Option<String>,
        #[arg(long)]
        to: Option<String>,
        #[arg(long, value_name = "SEMVER")]
        new_version: Option<String>,
        #[arg(long)]
        no_authors: bool,
        #[arg(long, value_name = "NAME_OR_EMAIL")]
        exclude_author: Vec<String>,
        #[arg(long)]
        hide_author_email: bool,
        #[arg(long)]
        clean: bool,
        #[arg(long)]
        sign: bool,
        /// Auto-confirm (skip prompts)
        #[arg(long)]
        yes: bool,
    },
    /// Run full release (bump, changelog, tag creation optional in future)
    Release {
        #[arg(long)]
        dry_run: bool,
        #[arg(long)]
        from: Option<String>,
        #[arg(long)]
        to: Option<String>,
        #[arg(long, value_name = "SEMVER")]
        new_version: Option<String>,
        #[arg(long)]
        no_authors: bool,
        #[arg(long, value_name = "NAME_OR_EMAIL")]
        exclude_author: Vec<String>,
        #[arg(long)]
        hide_author_email: bool,
        #[arg(long)]
        clean: bool,
        #[arg(long)]
        sign: bool,
        #[arg(long)]
        yes: bool,
    },
    /// GitHub release synchronization only
    Github {
        #[arg(long)]
        tag: String,
        #[arg(long)]
        body_path: Option<String>,
    },
}

pub fn run() -> Result<ExitCode> {
    let cli = Cli::parse();
    logging::init(cli.verbose as usize);
    let cwd = cli
        .cwd
        .as_ref()
        .map(std::path::PathBuf::from)
        .unwrap_or(std::env::current_dir()?);
    let exit = match cli.command {
        Commands::Show {
            from,
            to,
            new_version,
        } => {
            let parsed_new = new_version.and_then(|s| semver::Version::parse(&s).ok());
            let outcome = run_release(ReleaseOptions {
                cwd,
                from,
                to,
                dry_run: true,
                new_version: parsed_new,
                no_authors: true,
                exclude_authors: vec![],
                hide_author_email: false,
                clean: false,
                sign: false,
                yes: true, // Show command doesn't need confirmation
            })?;
            println!("{}", outcome.version);
            ExitCode::Success
        }
        Commands::Generate {
            write,
            output,
            from,
            to,
            new_version,
            no_authors,
            exclude_author,
            hide_author_email,
            clean,
            sign,
            yes,
        } => {
            let parsed_new = new_version.and_then(|s| semver::Version::parse(&s).ok());
            let outcome = run_release(ReleaseOptions {
                cwd: cwd.clone(),
                from,
                to,
                dry_run: !write,
                new_version: parsed_new,
                no_authors,
                exclude_authors: exclude_author,
                hide_author_email,
                clean,
                sign,
                yes,
            })?;
            if let Some(path) = output {
                std::fs::write(&path, outcome.version.to_string())?;
            }
            println!(
                "Generated v{} ({} commits){}",
                outcome.version,
                outcome.commit_count,
                if write {
                    if outcome.wrote {
                        " and updated CHANGELOG.md"
                    } else {
                        " (no change)"
                    }
                } else {
                    ""
                }
            );
            if !outcome.wrote && write {
                ExitCode::NoChange
            } else {
                ExitCode::Success
            }
        }
        Commands::Release {
            dry_run,
            from,
            to,
            new_version,
            no_authors,
            exclude_author,
            hide_author_email,
            clean,
            sign,
            yes,
        } => {
            let parsed_new = new_version.and_then(|s| semver::Version::parse(&s).ok());
            let outcome = run_release(ReleaseOptions {
                cwd: cwd.clone(),
                from,
                to,
                dry_run,
                new_version: parsed_new,
                no_authors,
                exclude_authors: exclude_author,
                hide_author_email,
                clean,
                sign,
                yes,
            })?;
            if outcome.wrote {
                println!("Released v{}", outcome.version);
                ExitCode::Success
            } else {
                println!("No change for v{}", outcome.version);
                ExitCode::NoChange
            }
        }
        Commands::Github { tag, body_path } => {
            // Minimal body read
            let body = if let Some(path) = body_path {
                std::fs::read_to_string(path)?
            } else {
                String::new()
            };
            // attempt repo detection via config layer
            let cfg = crate::config::load_config(crate::config::LoadOptions {
                cwd: &cwd,
                cli_overrides: None,
            })?;
            if let Some(repo) = cfg.repo {
                let rt = tokio::runtime::Runtime::new()?;
                let info = rt.block_on(async move {
                    github::sync_release(&repo, cfg.github_token.as_deref(), &tag, &body).await
                });
                match info {
                    Ok(r) => {
                        println!(
                            "GitHub release {}: {} (created={}, updated={}, skipped={})",
                            r.tag, r.url, r.created, r.updated, r.skipped
                        );
                        ExitCode::Success
                    }
                    Err(e) => {
                        eprintln!("github sync error: {e}");
                        ExitCode::NoChange
                    }
                }
            } else {
                eprintln!("no repository info available");
                ExitCode::NoChange
            }
        }
    };
    Ok(exit)
}
#[cfg(test)]
mod tests {
    use super::*;
    use clap::CommandFactory;
    #[test]
    fn cli_help_generates() {
        Cli::command().debug_assert();
    }
}
