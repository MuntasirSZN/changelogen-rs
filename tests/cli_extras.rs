use changelogen::cli::Cli;
use clap::Parser;

#[test]
fn unknown_subcommand_error() {
    let res = Cli::try_parse_from(["changelogen", "bogus"]);
    assert!(res.is_err());
}

#[test]
fn help_snapshot() {
    use clap::CommandFactory;
    let mut cmd = Cli::command();
    let help = cmd.render_long_help().to_string();
    insta::assert_snapshot!("cli_help", help);
}
