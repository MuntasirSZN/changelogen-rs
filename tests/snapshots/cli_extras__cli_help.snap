---
source: tests/cli_extras.rs
expression: help
---
changelogen 
Generate changelogs from conventional commits.

Usage: changelogen <COMMAND>

Commands:
  show      Show the next inferred version
  generate  Generate release block (writes with --write)
  release   Run full release (bump, changelog, tag creation optional in future)
  github    GitHub release synchronization only
  help      Print this message or the help of the given subcommand(s)

Options:
  --cwd <CWD>          
  -v...                Increase verbosity (-v, -vv, -vvv)
  -h, --help           Print help
  -V, --version        Print version
